### DISCUSSION TITLE
where are metrics calculated? in which python file

### DISCUSSION THREAD

**Discussion by CLeaf1023** [open-edge-platform/anomalib#418] [Q&A]
after       anomalib fit --config ./configs/model/patchcore.yaml
i got 
![图片](https://user-images.githubusercontent.com/31176427/176807410-04f72d61-3389-4220-89f6-330816dc5702.png)

in config file,i see
metrics:
  image_metrics:
    - F1Score
    - AUROC
  pixel_metrics:
    - F1Score
    - AUROC
  
plz help me find
where are metrics calculated? in which python file

**Comment by samet-akcay** [comment]
@cyj95, when the training is triggered, [`MetricConfigurationCallback`](https://github.com/openvinotoolkit/anomalib/blob/development/anomalib/utils/callbacks/metrics_configuration.py) is initialized. This creates and sets the image and pixel metrics to the anomaly module. Since this is a standard procedure for all the models in this library, we store these attributes in the base [`AnomalyModule`](https://github.com/openvinotoolkit/anomalib/blob/development/anomalib/models/components/base/anomaly_module.py). 

The metrics are then computed in [`validation_epoch_end`](https://github.com/openvinotoolkit/anomalib/blob/development/anomalib/models/components/base/anomaly_module.py#L125) method of the same module. Hope this helps. Let me know if you have any other questions.

**Reply by CLeaf1023** [reply]
I found “anomalib/utils/metrics/auroc.py” file，but I didn't find where AUROC is used?
And which file displays this result? 
![图片](https://user-images.githubusercontent.com/31176427/178220220-fba566e9-98c5-413e-a71d-0c8f57f2db3c.png)

**Reply by ashwinvaidya17** [reply]
For the second question, the metrics object is logged in `_log_metrics` function in the base anomaly module `anomalib/modules/components/base/anomaly_module.py`. PyTorch Lightning is responsible for displaying the metrics logged using `self.log_dict`. Maybe this might help https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#log-dict
